raiden/smart_contracts/TokenNetwork.sol

```diff
@@ -0,0 +1,568 @@
+        computed_locksroot = computeMerkleRoot(locked_encoded, merkle_proof);
+
+        // Note that unlocked locks have to be re-unlocked after a `transferUpdate` with a
+        // locksroot that does not contain this lock.
+        require(partner_state.locksroot == computed_locksroot);
+
+        // A lock can be withdrawn only once per participant
+        key = keccak256(partner_state.nonce, partner_state.locksroot);
+        require(!partner_state.withdrawn_locks[key]);
+        partner_state.withdrawn_locks[key] = true;
+
+        // Finally change the amount of owed tokens
+        // This implementation allows for each transfer to be set only once, so
+        // it's safe to update the transferred_amount in place.
+        //
+        partner_state.transferred_amount += amount;
+
+        ChannelUnlocked(channel_identifier, partner, partner_state.transferred_amount);
+    }
+
+    /// @notice Settles the balance between the two parties
+    function settleChannel(
+        uint channel_identifier,
+        address participant1,
+        address participant2)
+        isClosed(channel_identifier)
+        timeoutOver(channel_identifier)
+        public
+    {
+        uint participant1_amount;
+        uint participant2_amount;
+        uint total_deposit;
+
+        Participant memory participant1_state = channels[channel_identifier].participants[participant1];
+        Participant memory participant2_state = channels[channel_identifier].participants[participant2];
+
+        // Make sure the addresses are channel participant addresses
+        require(participant1_state.initialized);
+        require(participant2_state.initialized);
+
+        // Direct token transfers done through the token `transfer` function
+        // cannot be accounted for, these superfluous tokens will be burned,
+        // this is because there is no way to tell which participant (if any)
+        // had ownership over the token.
+        total_deposit = participant1_state.balance + participant2_state.balance;
+
+        participant1_amount = (
+            participant1_state.balance
+            + participant2_state.transferred_amount
+            - participant1_state.transferred_amount
+        );
+
+        // To account for cases when participant2 does not provide participant1's balance proof
+        // Therefore, participant1's transferred_amount will be lower than in reality
+        participant1_amount = min(participant1_amount, total_deposit);
+
+        // To account for cases when participant1 does not provide participant2's balance proof
+        // Therefore, participant2's transferred_amount will be lower than in reality
+        participant1_amount = max(participant1_amount, 0);
+
+        // At this point `participant1_amount` is between [0,total_deposit], so this is safe.
+        participant2_amount = total_deposit - participant1_amount;
+
+        // Remove the channel data from storage
+        delete channels[channel_identifier];
```

Comment:
I'm curious, does this operator works with the `mapping(address => Participant) participants;` in the channel?

Reply:
I don't see why it wouldn't

Reply:
how does it knows which keys are being used to actually free the storage?

Reply:
It should. I will say for sure after I test it. Until then, this is more of a draft of what should happen.